### PR TITLE
Feature/#100 enroll in course

### PR DESCRIPTION
**Pull request by m0e33**
This Pull Request implements:
#94 
#100 
#89 
#95 
#88 
#87 
#86

### CODE REVIEW

File: app/views/home/index.html.erb [hunk 1]
```diff
@@ -7,5 +7,48 @@
     <%# fa_icon helper: https://github.com/bokmann/font-awesome-rails %>
     <%= link_to (fa_icon 'sign-in', text: 'Log-In'), new_user_session_path, class: 'btn btn-primary btn-lg' %>
     <%= link_to (fa_icon 'user-plus', text: 'Sign-Up'), new_user_registration_path, class: 'btn btn-secondary btn-lg' %>
+    <% if user_signed_in? && !@current_user.is_student %>
+      <%= link_to (fa_icon 'graduation-cap', text: 'New Course'), new_course_path, class: 'btn btn-secondary btn-lg' %>
+    <br>
+    <p class="lead">My Courses</p>
```
Comment (phoeinx) [inline app/views/home/index.html.erb]: Should be indented

File: app/views/home/index.html.erb [hunk 2]
```diff
@@ -7,5 +7,48 @@
...
+        <thead>
+        <tr>
+          <th scope="col">Name</th>
+        </tr>
+        </thead>
+        <% @open_courses.each do |course| %>
+
+          <tr>
+            <td><%= course.name %></td>
+            <td><%#= button_to "Join", {action: "join_course", params: {id: course.id}}, {class:"btn btn-primary"} %></td>
```
Comment (phoeinx) [inline app/views/home/index.html.erb]: should be deleted

Comment (phoeinx) [inline app/views/home/index.html.erb]:
```suggestion
```

File: app/views/polls/index.html.erb [hunk 3]
```diff
@@ -17,15 +17,18 @@
         <td><%= poll.title %></td>
         <td><%= poll.is_multiselect %></td>
         <td><%= poll.is_active %></td>
-        <td><%= link_to 'Show', lecture_poll_path(@lecture, poll) %></td>
-        <td><%= link_to 'Edit', edit_lecture_poll_path(@lecture, poll) %></td>
-        <td><%= link_to 'Destroy', [@lecture, poll], method: :delete, data: { confirm: 'Are you sure?' } %></td>
+        <td><%= link_to 'Show', course_lecture_poll_path(course_id:@lecture.course.id, lecture_id:@lecture.id, id: poll.id) %></td>
+        <td><%= link_to 'Edit', edit_course_lecture_poll_path(course_id:@lecture.course.id, lecture_id:@lecture.id, id: poll.id) %></td>
+<!--        <td><%#= link_to 'Destroy', [@lecture, poll], method: :delete, data: { confirm: 'Are you sure?' } %></td>-->
```
Comment (phoeinx) [inline app/views/polls/index.html.erb]:
```suggestion
```

File: spec/controllers/courses_controller_spec.rb [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+    it "redirects to the courses list" do
+      course = Course.create! valid_attributes
+      delete :destroy, params: { id: course.to_param }, session: valid_session
+      expect(response).to redirect_to(courses_url)
+    end
+  end
+
+  describe "POST #join_course" do
+    before(:each) do
+      # login user
```
Comment (phoeinx) [inline spec/controllers/courses_controller_spec.rb]:
```suggestion
```

File: spec/controllers/courses_controller_spec.rb [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+    context "with valid params" do
+      let(:new_attributes) {
+        skip("Add a hash of attributes valid for your model")
+      }
+
+      it "updates the requested course" do
+        course = Course.create! valid_attributes
+        put :update, params: { id: course.to_param, course: new_attributes }, session: valid_session
+        course.reload
+        skip("Add assertions for updated state")
```
Comment (phoeinx) [inline spec/controllers/courses_controller_spec.rb]: Should be updated according to comment to actually test

File: spec/controllers/lectures_controller_spec.rb [hunk 6]
```diff
@@ -38,85 +41,94 @@
     end
   end
 
+  # there is no lectures index anymore because lectures only exists in the context of a course
```
Comment (phoeinx) [inline spec/controllers/lectures_controller_spec.rb]: delete uncommented code

File: spec/controllers/lectures_controller_spec.rb [hunk 7]
```diff
@@ -143,21 +155,21 @@
...
-      it "redirects for other lecturers", :logged_lecturer do
-        lecture = Lecture.create! valid_attributes_with_lecturer
-        put :update, params: { id: lecture.to_param, lecture: invalid_attributes }, session: valid_session
-        expect(response).to redirect_to(lectures_url)
+      it "redirects to root for other lecturers", :logged_lecturer do
+        lecture = Lecture.create! valid_attributes_with_lecturer_with_course
+        put :update, params: {course_id: lecture.course.id, id: lecture.to_param, lecture: invalid_attributes }, session: valid_session
+        expect(response).to redirect_to(root_path)
       end
       it "returns a success response for lecturer (i.e. to display the 'edit' template)", :logged_lecturer do
```
Comment (phoeinx) [inline spec/controllers/lectures_controller_spec.rb]: add test student cannot create lecture

File: spec/controllers/lectures_controller_spec.rb [hunk 8]
```diff
@@ -128,13 +140,13 @@
         { name: "SWT2", enrollment_key: "epic", status: "running" }
       }
       before(:each) do
-        @lecture = Lecture.create! valid_attributes_with_lecturer
+        @lecture = Lecture.create! valid_attributes_with_lecturer_with_course
         # login lecturer
```
Comment (phoeinx) [inline spec/controllers/lectures_controller_spec.rb]:
```suggestion
```

File: spec/factories/lectures.rb [hunk 9]
```diff
@@ -1,7 +1,8 @@
 FactoryBot.define do
   factory :lecture do
-    name { "SWTII" }
+    name { "SWT2" }
     enrollment_key { "ruby" }
     lecturer { FactoryBot.create(:user, :lecturer) }
+    course { FactoryBot.create(:course)}
```
Comment (phoeinx) [inline spec/factories/lectures.rb]:
```suggestion
    course { FactoryBot.create(:course) }
```

File: spec/views/courses/show.html.erb_spec.rb [hunk 10]
```diff
@@ -0,0 +1,23 @@
+require "rails_helper"
+
+RSpec.describe "courses/show", type: :view do
+  before(:each) do
+    @course = FactoryBot.create(:course)
+    @lecture = FactoryBot.create(:lecture, name: "Name", enrollment_key: "Enrollment", status: "created",  course: @course)
```
Comment (LucPrestin) [inline spec/views/courses/show.html.erb_spec.rb]:
```suggestion
    @lecture = FactoryBot.create(:lecture,  course: @course)
```

File: spec/views/courses/show.html.erb_spec.rb [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+  it "renders attributes of the course in <p>" do
+    render
+    expect(rendered).to match(/SWT2/)
+    expect(rendered).to match(/ruby/)
+    expect(rendered).to match(/open/)
+  end
+
+  it "renders a list of lectures" do
+    render
+    assert_select "tr>td", text: "Name".to_s, count: 1
```
Comment (LucPrestin) [inline spec/views/courses/show.html.erb_spec.rb]:
```suggestion
    assert_select "tr>td", text: @lecture.name, count: 1
```

File: spec/views/courses/show.html.erb_spec.rb [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+    assert_select "tr>td", text: "Enrollment".to_s, count: 1
```
Comment (LucPrestin) [inline spec/views/courses/show.html.erb_spec.rb]:
```suggestion
    assert_select "tr>td", text: @lecture.enrollment_key, count: 1
```

File: spec/views/courses/show.html.erb_spec.rb [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+    assert_select "tr>td", text: "created".to_s, count: 1
```
Comment (LucPrestin) [inline spec/views/courses/show.html.erb_spec.rb]:
```suggestion
    assert_select "tr>td", text: @lecture.status, count: 1
```

File: spec/views/home/index.html.erb_spec.rb [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+    expect(page).to have_link "New Course", href: new_course_path
+  end
+
+  it "displays all open courses for a student" do
+    FactoryBot.create(:course, name: "index_test")
+    FactoryBot.create(:course, name: "index_test")
+    @current_user = FactoryBot.create(:user, :student)
+    sign_in @current_user
+    visit root_path
+    expect(page).to have_selector("td", text: "index_test", count: 2)
```
Comment (LucPrestin) [inline spec/views/home/index.html.erb_spec.rb]: Ahhhh ein hardcoded String. Rette sich, wer kann.

File: spec/views/lectures/current.html.erb_spec.rb [hunk 15]
```diff
@@ -3,30 +3,39 @@
 RSpec.describe "lectures/current", type: :view do
   before(:each) do
     @lecturer = FactoryBot.create(:user, :lecturer, email: "test@test.de")
+    @course = FactoryBot.create(:course)
     @lectures = assign(:lectures, [
```
Comment (LucPrestin) [inline spec/views/lectures/current.html.erb_spec.rb]: Use the FactoryBot for this

File: spec/views/lectures/current.html.erb_spec.rb [hunk 16]
```diff
@@ -3,30 +3,39 @@
...
+        course: @course
       )
     ])
   end
 
-  it "renders a list of lectures" do
-    render
-    expect(rendered).to have_css("td", text: @lectures[0].name)
-    expect(rendered).to have_css("td", text: @lectures[1].name)
+  it "renders a list of aöll running lectures" do
```
Comment (LucPrestin) [inline spec/views/lectures/current.html.erb_spec.rb]:
```suggestion
  it "renders a list of all running lectures" do
```

File: spec/views/lectures/edit.html.erb_spec.rb [hunk 17]
```diff
@@ -2,18 +2,20 @@
 
 RSpec.describe "lectures/edit", type: :view do
   before(:each) do
+    @course = FactoryBot.create(:course)
     @lecture = assign(:lecture, Lecture.create!(
                                   name: "MyString",
                                   enrollment_key: "MyString",
                                   status: "created",
-                                  lecturer: FactoryBot.create(:user, :lecturer, email: "123test@gmail.com")
+                                  lecturer: FactoryBot.create(:user, :lecturer, email: "123test@gmail.com"),
```
Comment (LucPrestin) [inline spec/views/lectures/edit.html.erb_spec.rb]:
```suggestion
                                  lecturer: FactoryBot.create(:user, :lecturer),
```

File: spec/views/lectures/index.html.erb_spec.rb [hunk 18]
```diff
@@ -2,26 +2,32 @@
...
 RSpec.describe "lectures/index", type: :view do
   before(:each) do
+    @course = FactoryBot.create(:course)
     assign(:lectures, [
       Lecture.create!(
         name: "Name",
         enrollment_key: "Enrollment Key",
         status: "created",
-        lecturer: FactoryBot.create(:user, :lecturer, email: "hp@hpi.de")
+        lecturer: FactoryBot.create(:user, :lecturer, email: "hp@hpi.de"),
```
Comment (LucPrestin) [inline spec/views/lectures/index.html.erb_spec.rb]:
```suggestion
        lecturer: FactoryBot.create(:user, :lecturer),
```

File: spec/views/lectures/index.html.erb_spec.rb [hunk 19]
```diff
@@ -2,26 +2,32 @@
...
+        course:@course
       ),
       Lecture.create!(
         name: "Name",
         enrollment_key: "Enrollment Key",
         status: "created",
-        lecturer: FactoryBot.create(:user, :lecturer, email: "cm@hpi.de")
+        lecturer: FactoryBot.create(:user, :lecturer, email: "cm@hpi.de"),
```
Comment (LucPrestin) [inline spec/views/lectures/index.html.erb_spec.rb]:
```suggestion
        lecturer: FactoryBot.create(:user, :lecturer),
```

File: spec/views/lectures/index.html.erb_spec.rb [hunk 20]
```diff
@@ -2,26 +2,32 @@
...
+        course:@course
       )
     ])
   end
 
+# this test is now in courses/show.html.erb_spec due that lectures don't have an own index anymore
```
Comment (LucPrestin) [inline spec/views/lectures/index.html.erb_spec.rb]: Then delete it here!

File: spec/views/lectures/new.html.erb_spec.rb [hunk 21]
```diff
@@ -2,17 +2,14 @@
 
 RSpec.describe "lectures/new", type: :view do
   before(:each) do
-    assign(:lecture, Lecture.new(
-                       name: "MyString",
-                       enrollment_key: "MyString",
-                       status: "created"
-    ))
+    @course = FactoryBot.create(:course)
+    @lecture = FactoryBot.create(:lecture, name:"MyString",enrollment_key: "MyString",status: "created",course: @course)
```
Comment (LucPrestin) [inline spec/views/lectures/new.html.erb_spec.rb]:
```suggestion
    @lecture = FactoryBot.create(:lecture, course: @course)
```

File: spec/views/lectures/show.html.erb_spec.rb [hunk 22]
```diff
@@ -3,12 +3,14 @@
...
   let(:valid_session) { {} }
   before(:each) do
+    @course = FactoryBot.create(:course)
     @lecture = assign(:lecture, Lecture.create!(
                                   name: "Name",
                                   description: "Test",
                                   enrollment_key: "Enrollment Key",
                                   status: "created",
-                                  lecturer: FactoryBot.create(:user, :lecturer, email: "bp@hpi.de")
+                                  lecturer: FactoryBot.create(:user, :lecturer, email: "bp@hpi.de"),
```
Comment (LucPrestin) [inline spec/views/lectures/show.html.erb_spec.rb]:
```suggestion
                                  lecturer: FactoryBot.create(:user, :lecturer),
```

File: spec/views/lectures/show.html.erb_spec.rb [hunk 23]
```diff
@@ -3,12 +3,14 @@
 RSpec.describe "lectures/show", type: :view do
   let(:valid_session) { {} }
   before(:each) do
+    @course = FactoryBot.create(:course)
     @lecture = assign(:lecture, Lecture.create!(
```
Comment (LucPrestin) [inline spec/views/lectures/show.html.erb_spec.rb]: Could we use the FactoryBot for this?

**Comment by phoeinx** [comment]
Included in pull request for #92.

**Pull request closed without merging**